### PR TITLE
feat(sql): add trigonometric functions sin, cos, tan, cot, asin, acos, atan, atan2, toRadians, fromRadians, pi

### PR DESCRIPTION
**Pull request by marregui**
These functions are useful for trigonometry: sin, cos, tan, toRadians, fromRadians, pi, acos, asin, atan, atan2, cot.

toRadians and  fromRadians are complementary.
pi is the constant.
sin, cos, tan, cot take the angle in radians as their parameter.
atan2 takes two doubles.
atan, asin, acos take one double.

documented by https://github.com/questdb/questdb.io/pull/1245

### CODE REVIEW

**Comment by puzpuzpuz** [comment]
#2842 seems to be relevant, so it's worth adding a reference

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 123 / 125 (98.40%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/Atan2DoubleFunctionFactory.java|14|16|87.50%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactory.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AsinDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/TanDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AcosDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/PIDoubleFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/CotDoubleFunctionFactory.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/SinDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/FromRadiansDoubleFunctionFactory.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/CosDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AtanDoubleFunctionFactory.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 125 / 127 (98.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/Atan2DoubleFunctionFactory.java|14|16|87.50%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactory.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AsinDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/TanDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AcosDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/PIDoubleFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/CotDoubleFunctionFactory.java|17|17|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/SinDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/FromRadiansDoubleFunctionFactory.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/CosDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AtanDoubleFunctionFactory.java|11|11|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import io.questdb.std.IntList;
+import io.questdb.std.ObjList;
+
+public class ToRadiansDoubleFunctionFactory implements FunctionFactory {
+
+    private static final double TO_RAD = Math.PI / 180.0;
+
+    @Override
+    public String getSignature() {
+        return "toRadians(D)";
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactory.java:44]: Have you considered "radians(D)" for function name?
Just to be "closer" to the PostgreSQL.

File: core/src/main/java/io/questdb/griffin/engine/functions/math/FromRadiansDoubleFunctionFactory.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import io.questdb.std.IntList;
+import io.questdb.std.ObjList;
+
+public class FromRadiansDoubleFunctionFactory implements FunctionFactory {
+
+    private static final double FROM_RAD = 180.0 / Math.PI;
+
+    @Override
+    public String getSignature() {
+        return "fromRadians(D)";
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/FromRadiansDoubleFunctionFactory.java:44]: Same here: PostgreSQL uses "degrees(D)"

File: core/src/main/java/io/questdb/griffin/engine/functions/math/AcosDoubleFunctionFactory.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            int position, ObjList<Function> args,
+            IntList argPositions,
+            CairoConfiguration configuration,
+            SqlExecutionContext sqlExecutionContext
+    ) {
+        Function x = args.getQuick(0);
+        if (x.isConstant()) {
+            return new DoubleConstant(StrictMath.acos(x.getDouble(null)));
+        }
+        return new AcosFunction(args.getQuick(0));
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/AcosDoubleFunctionFactory.java:56]: x can be reused here.

File: core/src/main/java/io/questdb/griffin/engine/functions/math/AsinDoubleFunctionFactory.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public Function newInstance(
+            int position, ObjList<Function> args,
+            IntList argPositions,
+            CairoConfiguration configuration,
+            SqlExecutionContext sqlExecutionContext
+    ) {
+        Function x = args.getQuick(0);
+        if (x.isConstant()) {
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/AsinDoubleFunctionFactory.java:53]: We don't have such const optimisation for other math function. Does it improve anything?

File: core/src/main/java/io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactory.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    ) {
+        Function angleDecimal = args.getQuick(0);
+        if (angleDecimal.isConstant()) {
+            return new DoubleConstant(toRadians(angleDecimal.getDouble(null)));
+        }
+        return new ToRadians(args.getQuick(0));
+    }
+
+    private static double toRadians(double angle) {
+        return angle != angle ? Double.NaN : (angle % 360) * TO_RAD;
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactory.java:62]: Can we use Math.toRadians() here?

File: core/src/main/java/io/questdb/griffin/engine/functions/math/FromRadiansDoubleFunctionFactory.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    ) {
+        Function angleRad = args.getQuick(0);
+        if (angleRad.isConstant()) {
+            return new DoubleConstant(fromRadians(angleRad.getDouble(null)));
+        }
+        return new FromRadians(args.getQuick(0));
+    }
+
+    private static double fromRadians(double angle) {
+        return angle != angle ? Double.NaN : (angle % 360) * FROM_RAD;
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/FromRadiansDoubleFunctionFactory.java:62]: Can we use Math.toDegrees() here?

File: core/src/test/java/io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.engine.AbstractFunctionFactoryTest;
+import org.junit.Test;
+
+public class ToRadiansDoubleFunctionFactoryTest extends AbstractFunctionFactoryTest {
+
+    @Test
+    public void testNaN() throws SqlException {
+        call(Double.NaN).andAssert(Double.NaN, 0.000000000000001);
```
Comment (eugenels) [inline core/src/test/java/io/questdb/griffin/engine/functions/math/ToRadiansDoubleFunctionFactoryTest.java:36]: Same, DELTA

File: core/src/main/java/io/questdb/griffin/engine/functions/math/CotDoubleFunctionFactory.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            SqlExecutionContext sqlExecutionContext
+    ) {
+        Function angleRad = args.getQuick(0);
+        if (angleRad.isConstant()) {
+            return new DoubleConstant(cot(angleRad.getDouble(null)));
+        }
+        return new CotFunction(args.getQuick(0));
+    }
+
+    private static double cot(double angle) {
```
Comment (eugenels) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/CotDoubleFunctionFactory.java:59]: The result does not match the PostgreSQL's cot(). I'm waiting for user complaints.

File: core/src/main/java/io/questdb/griffin/engine/functions/math/AsinDoubleFunctionFactory.java (hunk 4 above)
Comment (marregui) [reply core/src/main/java/io/questdb/griffin/engine/functions/math/AsinDoubleFunctionFactory.java]: nope, removing

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 93 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AsinDoubleFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/TanDoubleFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AcosDoubleFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/PIDoubleFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/CotDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/Atan2DoubleFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/SinDoubleFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/RadiansDoubleFunctionFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/CosDoubleFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/AtanDoubleFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/DegreesDoubleFunctionFactory.java|11|11|100.00%|

**Review by eugenels — APPROVED** [review]

**Pull request merged**